[wongkj12] iP

wongkj12 opened the pull request:
# Duke

> "Hey man, i'm MR. WONG. How may I help you today?" - Duke

**Duke** is a task manager chatbot that can help you to
- Remember todo lists
- Set deadlines
- Keep track of events
- and many more!

All you need to do is,
1. Download it from [here](https://github.com/wongkj12/ip/releases)
2. Run it
3. Add your tasks using _text-based_ commands
4. Let it manage your tasks! 🗿  

Current features:
- [x] Managing todos
- [x] Deadlines
- [x] Events
- [ ] Reminders (maybe coming soon)

Here is the current `main` method:
```java
public class Duke {
    public static void Duke(String[] args) {
        // ...
        }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

wongkj12 committed:
Rename, Greet, Exit

wongkj12 committed:
Echo

wongkj12 committed:
Add, List

wongkj12 committed:
Added Mark, Todo, Event, Deadline functionality. Also made Parser class to parse commands

wongkj12 committed:
Automated text UI testing

wongkj12 committed:
added simple exceptions and error handling

wongkj12 committed:
Delete feature

wongkj12 committed:
Add writeToFile

wongkj12 committed:
Update load data functionality

wongkj12 committed:
Merge branch 'branch-Level-7'

wongkj12 committed:
Update datetime functionality

wongkj12 committed:
Merge branch 'branch-Level-8'

wongkj12 committed:
Update OOP

wongkj12 committed:
Update packages

wongkj12 committed:
Merge branch 'add-gradle-support'

wongkj12 committed:
Add gradle

wongkj12 committed:
Add parser JUnit test

wongkj12 committed:
Add deadline JUnit test

wongkj12 committed:
Update javadocs

wongkj12 committed:
Add find command

wongkj12 committed:
Merge branch 'A-Javadoc'

Wongqingyee commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,84 @@
+package duke;
+
+import duke.command.*;
```
Avoid using * wildcard.

dedsecrattle replied:
Can we avoid widlcard imports in order to adhere to the coding style ?

Wongqingyee commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,144 @@
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.io.File;
```
Ordering of the import statements should be consistent. 
It should be arranging all the java.io. first, then followed by java.util.

Wongqingyee commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,144 @@
+ * TaskList class to represent the collection of tasks being maintained.
+ */
+public class TaskList {
+    private List<Task> list;
```
Plural form should be used on names representing a collection of objects.

Wongqingyee commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,28 @@
+package duke;
+
+import org.junit.jupiter.api.Test;
```
The import statements should be in the following order:
import static ...
<blank line>
import ...

Wongqingyee commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,19 @@
+package duke.task;
+
+import org.junit.jupiter.api.Test;
```
Same issue for the ordering of the import statements.

dedsecrattle commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,19 @@
+        this.index = index;
+    }
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
```
Can we add javadoc for the method `execute`?

dedsecrattle commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,52 @@
+            fromDate = LocalDate.parse(from);
+            this.from = fromDate.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+
+        } catch (DateTimeParseException e) {}
```
Can we try to convey the user about this `DateTimeParseException` instead of just catching it ?

wongkj12 committed:
Add JavaFX GUI, remove UI class

wongkj12 committed:
Merge branch 'branch-GUI'

jovantanyk commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,12 @@
+    public boolean isExit() {
+        return this.isExit;
+    }
+    public abstract String execute(TaskList tasks, Storage storage);
```
Could use higher level of abstraction here. Maybe have another class to handle UI related methods instead of returning String directly as there might be commands that dont require any output.

jovantanyk commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+     * @param name Name of the deadline
+     * @param by Due date of the deadline
+     */
+    public Deadline(String name, String by) {
+        super(name);
+        this.by = by;
+
+        // Check if they are in dateTime format
+        try {
+            byDate = LocalDate.parse(by);
+            this.by = byDate.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+        } catch (DateTimeParseException e) {}
+    }
```
Try and avoid having too much business logic in class constructors. The input should already be parsed and processed by the Parser.java class instead of letting Deadline.java handle it

jovantanyk commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,52 @@
+package duke.task;
+
+/**
+ * Class representing a generic Task
+ */
+public class Task {
```
Since Task is never instantiated directly, you should make it an abstract class and allow subclasses to have their own implementation of the methods instead of overriding every method when Task is never instantiated to begin with

jovantanyk commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,55 @@
+     * @param filepath
+     */
+    public Storage(String filepath) {
+        this.filepath = filepath;
```
Since filepath will never change across the lifecycle of a program, it might be better to make Storage a static class instead, so that you won't have to pass an instance of Storage around every time you want to load or save a file

jovantanyk reviewed (commented):
Overall Good display of OOP knowledge and consistent method and class names.  Could use higher level of abstraction like creating a Ui class to handle output instead of having every Command deal with it directly as it creates a level of inconsistency.

wongkj12 committed:
Update javadocs

wongkj12 committed:
Add several assert statements

This commit uses the assert feature to document several important
assumptions that should hold at various points in the code.

If assertions are enabled, execution will terminate when index given
to mark commands are negative, for example.

wongkj12 committed:
Update comment

wongkj12 committed:
Merge pull request #3 from wongkj12/branch-A-Assertions

Add Assertions

wongkj12 committed:
Merge branch 'master' into branch-A-CodeQuality

wongkj12 committed:
Merge pull request #2 from wongkj12/branch-A-CodeQuality

Update code quality

wongkj12 committed:
Add tag feature

Added the ability to tag different tasks. The format
for tagging is as follows:

tag [index] [name]

and the corresponding tag will appear for the selected
task. Currently there is no way to remove tags, so
that could be implemented in the future.

wongkj12 committed:
Merge branch 'branch-C-Tagging'

wongkj12 committed:
Fix tagging bug

wongkj12 committed:
Add Ui.png

wongkj12 committed:
Add product page

wongkj12 committed:
Fix test case
